### PR TITLE
Closes #7869: Load a non-mobile url when desktop mode is toggled on

### LINKED ISSUE #7869: toggleDesktopMode to reload a non-mobile url when toggled on
For https://github.com/mozilla-mobile/fenix/issues/3903:

We should reload a non-mobile url when the desktop mode is toggled on; otherwise for certain mobile websites like youtube (`https://m.youtube.com`), we would get a redirect url: `https://www.youtube.com?app=desktop`.

### PR DESCRIPTION
**Pull request by psymoon**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -254,8 +259,44 @@ class GeckoEngineSession(
...
+
+        foundPrefix?.let {
+            val mobileUri = URI(
+                uri.scheme,
+                authority.substring(it.length),
+                uri.path,
+                uri.query,
+                uri.fragment
+            )
+            overrideUrl = mobileUri.toString()
```
Comment (NotWoods) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:296]: Could just return here and avoid the `var`

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 2]
```diff
@@ -254,8 +259,44 @@ class GeckoEngineSession(
...
+        val authority = uri.authority?.toLowerCase(Locale.ROOT) ?: return null
+
+        val foundPrefix = when {
+            authority.startsWith(mPrefix) -> mPrefix
+            authority.startsWith(mobilePrefix) -> mobilePrefix
+            else -> null
         }
+
+        foundPrefix?.let {
+            val mobileUri = URI(
```
Comment (NotWoods) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:289]: Can we use the Android `Uri` class instead, which is faster as it doesn't do any validation?

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (psymoon) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:296]: detekt `ReturnCount = 2` - https://arturbosch.github.io/detekt/style.html. Having said that, there may be better way to do this without using var 🤔 .

Comment (NotWoods) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:296]: If detekt is complaining about return count then `var` is fine

**Comment by psymoon** [comment]
bors r=NotWoods

**Pull request merged**